### PR TITLE
Support $CURRENT_ITEM in validations and presets

### PR DESCRIPTION
**Pull request by u12206050**
#WIP
## Description

CURRENT_ITEM

## Type of Change

- [ ] Bugfix
- [x] Improvement
- [x] New Feature
- [ ] Refactor / codestyle updates
- [ ] Other, please describe:

## Requirements Checklist

- [x] Tests are included
- [x] All tests are passing locally

If adding a new feature:

- [ ] Documentation was added/updated

### CODE REVIEW

**Comment by u12206050** [comment]
@rijkvanzanten you'll that this removes the need for checkAccess during update and create since we always try to fetch the items, together with any fields that we need for presets and validation. However, due to this I have refactored it so that not all fields need to be fetched, as a memory performance improvement. That is why I also had to update the tests in `items.test.ts`

**Comment by licitdev** [comment]
Let's keep track on whether this PR resolves the issues https://github.com/directus/directus/issues/14759 and https://github.com/directus/directus/issues/13198. Thanks!

**Comment by w0kyj** [comment]
Seems like this PR implements features https://github.com/directus/directus/discussions/2993

**Comment by AaronBeaudoin** [comment]
This PR appears to also add support for what I'm trying to do https://github.com/directus/directus/discussions/11905. Looking forward to seeing it merged!

**Comment by u12206050** [comment]
This PR isn't getting much attention it seems, even though it will solve a number of issues. @rijkvanzanten 
I my self am not sure what more I can do from my end, any ideas?

**Comment by rijkvanzanten** [comment]
> This PR isn't getting much attention it seems, even though it will solve a number of issues. @rijkvanzanten I my self am not sure what more I can do from my end, any ideas?

It's in the queue, but I can't give any guarantees or updates on timeline beyond that 👍🏻

**Comment by adanielyan** [comment]
Having this merged would be fantastic. Is there anything we can do to accelerate it?

**Comment by rijkvanzanten** [comment]
> Having this merged would be fantastic. Is there anything we can do to accelerate it?

We do offer [accelerated engineering](https://docs.directus.io/getting-started/support.html#sponsored-work), but other than that, we're reviewing/merging PRs according to the [contributing guidelines](https://github.com/directus/directus/blob/main/contributing.md) and internal prioritization 👍🏻

**Comment by arladmin** [comment]
Any idea regarding the plan for this PR to be merged?

**Comment by arladmin** [comment]
Any idea regarding the plan for this PR to be merged?

Until this feature is merged, have to resort to writing direct queries to the database, which beats the whole point of having this tool on top of the database.

File: api/src/services/authorization.ts [hunk 1]
```diff
@@ -554,52 +546,146 @@ export class AuthorizationService {
...
+			itemContexts = await this.getItemsContext(action, collection, pk, itemContextFields);
+
+			// Remove pk field so we don't check it during hashing later on
+			itemContextFields.splice(itemContextFields.indexOf(pkField), 1);
 		}
 
-		const validationErrors: FailedValidationException[] = [];
+		const chunks: Record<string, PayloadChunk> = {};
+		itemContexts.forEach((itemContext) => {
+			itemContext = merge({}, itemContext, payload);
```
Comment (jacobrienstra) [inline api/src/services/authorization.ts:607]: Hey, just looking over this as I was looking to implement something similar, and I think a potential issue you could run into here is the detailed syntax for related items, i.e.
```js
{ 
   create: [],
   update: [],
   delete: [],
}
```
showing up in the payload, and therefore not properly merging with the itemContext. Especially if there are some nested updates that need to be validated. I guess you'd need to merge updated items, add created items, and remove deleted items from the itemContext... recursively?

Comment (jacobrienstra) [reply api/src/services/authorization.ts:607]: Similar to this issue: https://github.com/directus/directus/pull/17842

**Comment by dahngeek** [comment]
willing to see this published :)

**Comment by OlivierPineau-qc** [comment]
Willing to see this being merged :) 
It's a great feature that I think will really help with advanced use of directus.

**Comment by Justin99b** [comment]
@u12206050 I this still up to date? I it possible to share a single commit with only neccessary changes? Would like to merge it into the current persion and see if it works :)

**Comment by Nisgrak** [comment]
I think this is super necessary for a minimal validation in the Access Control filters

**Comment by infomus** [comment]
This PR would solve a lot of problems for us! Any progress on it?

**Comment by JohnDebono** [comment]
@rijkvanzanten Are there any plans of having this merged in future releases?

**Comment by jorgekorgut** [comment]
__If someone in the future is looking to filter the payload based on the $CURRENT_ITEM, this comment could be usefull to you.__

It is a work around, but since this feature has fallen into backlogs, this was my best option to avoid furthering myself from the main branch.

The following code is a middleware implemented with Directus extension hooks that can filter permissions based on the $CURRENT_ITEM (extendedPayload).

The code is a little awkward and not at all generalized, but maybe this could save you some time.

**Main idea :**
Inside the Directus permissions tab, we let everything pass. Then we drop the permissions based on our logic inside a Directus extension.

If there is a problem with this idea, I would be glad to know ;)

How to create a new directus-extension :   
https://docs.directus.io/extensions/creating-extensions.html

How to install a directus-extension :  
https://docs.directus.io/extensions/installing-extensions.html

```ts
filter("Custom_collection.items.read", async (payload: any, eventMeta: any, context: any) => {

		//Skip filter if admin user
		if (context.accountability?.admin) {
			return payload;
		}

		const userId = context.accountability?.user;

		if (!userId) {
			console.error('No user found')
			return {}
		}

		const usersService = new UsersService({
			...context,
			ccountability: { admin: true },
			schema: schema
		});

		const itemsService = new ItemsService(
			eventMeta.collection,
			{
				...context,
				accountability: { admin: true },
				schema: schema
			});

		const queryUserInformation = {
			fields: ['id',
				'Settings_id.Feature',
			],
		}

		const queryItemInformation = {
			fields: ['id',
				'user_created.id',
				'user_created.Settings_id.Feature',],
		}

		try {
			const userContext = await usersService.readOne(userId, queryUserInformation);
			const itemKey = payload.map((item: any) => item.id);
			const extendedPayload = await itemsService.readMany(itemKey, queryItemInformation, { emitEvents: false })

			const toRemovePayloadId = extendedPayload.filter((item: any) => {

				const isFeatureNull = item.user_created.Settings_id?.Feature == null;
				const hasSameFeature = item.user_created.Settings_id?.Feature === userContext.Settings_id?.Feature;

				if (item.user_created.id === userContext.id ||
					(!isFeatureNull && hasSameFeature)
				) {
					return false;
				}
				return true;

			}).map((item: any) => item.id);

			const payloadFiltered = payload.filter((item: any) => {
				return !toRemovePayloadId.includes(item.id);
			})

			return payloadFiltered;

		} catch (error) {
			console.error(error)
		}
});
```

**Review by dmpinchuk — APPROVED** [review]

**Review by 2chiefk — APPROVED** [review]

**Comment by loongmxbt** [comment]
In my case, my directus_users have a company field (M2O) and a group field (M2O). Company (One to Many) Groups.
And I want the directus_users 'group' field only list the groups for the current company. Seems related to this feature.
image

**Comment by falmanna** [comment]
> In my case, my directus_users have a company field (M2O) and a group field (M2O). Company (One to Many) Groups. And I want the directus_users 'group' field only list the groups for the current company. Seems related to this feature. image

Tthis is doable without the `$CURRENT_ITEM` feature. 
You can simply do `Company Equals $CURRENT_USER.company_id`
More correctly, this should be implemented in the users `read` permission of the group collection and not in the interface section of the group item.

**Comment by adanielyan** [comment]
Good point, @falmanna   

However it's not always about access control. Sometimes you may want to filter based on a current item, access to which is not necessarily limited by permissions.

**Comment by falmanna** [comment]
> Good point, @falmanna
> 
> However it's not always about access control. Sometimes you may want to filter based on a current item, access to which is not necessarily limited by permissions.

Sure thing, but I was commenting on hist exact case, which seems like a mutli-tenancy setup. And for me this should be handled in permissions as interface is just a hiding mechanism that is not applicable to APIs as well.

**Comment by arladmin** [comment]
> In my case, my directus_users have a company field (M2O) and a group field (M2O). Company (One to Many) Groups. And I want the directus_users 'group' field only list the groups for the current company. Seems related to this feature. 

Try ->
`Company Equals {{company}}`

**Comment by falmanna** [comment]
> > In my case, my directus_users have a company field (M2O) and a group field (M2O). Company (One to Many) Groups. And I want the directus_users 'group' field only list the groups for the current company. Seems related to this feature.
> 
> Try -> `Company Equals {{company}}`

Is this syntax documented anywhere? just saw it today in another issue. Seems it is applicable to filters only.

**Comment by loongmxbt** [comment]
> > In my case, my directus_users have a company field (M2O) and a group field (M2O). Company (One to Many) Groups. And I want the directus_users 'group' field only list the groups for the current company. Seems related to this feature.
> 
> Try -> `Company Equals {{company}}`

This is exactly what I want! Can be very useful in other M2O filter scenarios. Can filter M2O options based on a field inside current item.

Thanks @falmanna. I intended to use admin account to set different users's company(different) and group(related to company). Your answer is suitable for the user has its own company and set its own group.

**Comment by PhungVanHoa** [comment]
any progress on this?

**Comment by rijkvanzanten** [comment]
Closing this as we're aiming to support this across all permissions rather than just validation/presets in a bigger sprint. Also there also seem to be some security questions around merging the user provided payload with the validation rules before validating, and the current iteration also isn't compatible with the policies-based permissions system, but those are solvable problems 🙂

**Pull request closed without merging**